hazelcast/src/test/java/com/hazelcast/test/jitter/JitterRule.java

```diff
@@ -0,0 +1,104 @@
+import static com.hazelcast.util.StringUtil.LINE_SEPARATOR;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+/**
+ * JUnit rule for detecting JVM/OS hiccups. It's meant to give you an insight into your environment
+ * in the case of a test failure. This is useful for troubleshooting of spuriously failing tests.
+ */
+public class JitterRule implements TestRule {
+    /**
+     * Time interval aggregated into a single bucket. Smaller interval provides
+     * a clearer picture about hiccups in time, too small intervals may use too
+     * much of memory and can also generate overwhelming amount of data.
+     */
+    public static final long AGGREGATION_INTERVAL_MILLIS = SECONDS.toMillis(5);
+
+    /**
+     * Number of buckets to be created. Jitter monitor records a floating window
+     * where the length of the window can be calculated as
+     * <code>AGGREGATION_INTERVAL_MILLIS * CAPACITY</code>
+     *
+     * It has to be a power of two.
+     */
+    public static final  int CAPACITY = nextPowerOfTwo(720);
+
+    /**
+     * Resolution of the measurement. Smaller number can detect shorter pauses,
+     * but it can cause too much of overhead. Too long value causes less overhead,
+     * but it may miss shorter pauses.
+     */
+    public static final long RESOLUTION_NANOS = MILLISECONDS.toNanos(10);
+
+    /**
+     * Hiccups over this threshold will be counted separately. This is useful for counting
+     * serious hiccups.
+     *
+     */
+    public static final long LONG_HICCUP_THRESHOLD = SECONDS.toNanos(1);
+
+    private static final String MODE_PROPERTY_NAME = "hazelcast.jitterMonitor.mode";
+    private static final Mode DEFAULT_MODE = Mode.JENKINS;
+    private static final boolean ENABLED = isEnabled();
+
+    static {
+        if (ENABLED) {
+            JitterMonitor.ensureRunning();
+        }
+    }
+
+    private static boolean isEnabled() {
+        String modePropertyValue = System.getProperty(MODE_PROPERTY_NAME, DEFAULT_MODE.name());
+        Mode mode = Mode.valueOf(modePropertyValue);
+        switch (mode) {
+            case DISABLED:
+                return false;
+            case ENABLED:
+                return true;
+            case JENKINS:
+                return JenkinsDetector.isOnJenkins();
+            default:
+                throw new IllegalArgumentException("Unknown mode: " + mode);
+        }
+    }
+
+    @Override
+    public Statement apply(final Statement base, final Description description) {
+        if (!ENABLED) {
+            return base;
+        }
+
+        return new Statement() {
+            @Override
+            public void evaluate() throws Throwable {
+                long startTime = System.currentTimeMillis();
+                try {
+                    base.evaluate();
+                } catch (Throwable t) {
+                    long endTime = System.currentTimeMillis();
```

Comment:
We should possibly log the jitter no matter if the test failed or succeed. A rule that is executed after the JitterRule can cause the test to fail, depending on the order in which rules are applied.